### PR TITLE
Fixes #147 - fix set_reference() with templates

### LINKED ISSUE #147: Various _set_reference() invocations don't work with templates
For example, `set_vsys()` does not work for Templates.

### PR DESCRIPTION
**Pull request by shinmog**
Few other updates:

* Adds a new flag, `return_boolean`, that the Ansible modules will need
* Fixes `set_vsys()` for VLANs, virtual wires, and virtual routers (only interface imports worked)

### CODE REVIEW

File: pandevice/base.py [hunk 1]
```diff
@@ -1367,18 +1367,73 @@ def _convert_var(cls, value, vartype):
...
         the zone to the interface. If the desired reference already exists then nothing happens.
 
         """
-        device = self.nearest_pandevice()
+        parent = None
+        update_needed = False
+
         if refresh:
-            allobjects = reference_type.refreshall(device, running_config=running_config)
+            from pandevice.firewall import Firewall
```
Comment (btorresgil) [inline pandevice/base.py]: Can you use comments here to document why this new refresh algorithm is necessary and, at a high level, what it does?

File: pandevice/base.py [hunk 2]
```diff
@@ -2934,27 +3001,37 @@ def set_vsys(self, vsys_id, refresh=False, update=False, running_config=False):
             update (bool): Apply the changes to the device (Default: False)
             running_config (bool): If refresh is True, refresh from the running
                 configuration (Default: False)
+            return_boolean (bool): Return a boolean saying if an update
+                is/was needed instead of a Vsys object.
```
Comment (btorresgil) [inline pandevice/base.py]: Expand on this to documentation to explain what a True and False value represents and how that relates to if it was updated or not.  Also consider a more descriptive parameter name.

File: pandevice/base.py [hunk 3]
```diff
@@ -1388,35 +1443,46 @@ def _set_reference(self, reference_name, reference_type, reference_var, exclusiv
...
                     if reference_name is not None and str(getattr(obj, reference_type.NAME)) == reference_name:
                         continue
+                    update_needed = True
                     references = None
                     if update: obj.update(reference_var)
+
         # Add new reference to self in requested object
         if reference_name is not None:
-            obj = device.find_or_create(reference_name, reference_type, *args, **kwargs)
+            obj = parent.find_or_create(reference_name, reference_type, **kwargs)
```
Comment (btorresgil) [inline pandevice/base.py]: Is `find_or_create` still appropriate when only checking if a change is needed?

Comment (shinmog) [reply pandevice/base.py]: It looks like `find_or_create()` doesn't actually do `.create()`.  It just adds the object to the object tree if it doesn't exist.  So no modifications to this are needed for `update=False`.

**Comment by shinmog** [comment]
I will be doing live tests, as mock tests don't bring enough certainty of no changes.

**Pull request merged**